Source/Tests/LinuxSyscalls/x32/Memory.cpp

```diff
@@ -19,6 +19,38 @@ tags: LinuxSyscalls|syscalls-x86-32
 #include <filesystem>
 
 namespace FEX::HLE::x32 {
+
+  void *x32SyscallHandler::GuestMmap(void *addr, size_t length, int prot, int flags, int fd, off_t offset) {
+    LOGMAN_THROW_A_FMT((length >> 32) == 0, "values must fit to 32 bits");
+
+    auto Result = (uint64_t)GetAllocator()->mmap((void*)addr, length, prot, flags, fd, offset);
+
+    LOGMAN_THROW_A_FMT((Result >> 32) == 0|| (Result >> 32) == 0xFFFFFFFF, "values must fit to 32 bits");
```

Comment:
`static_cast<int64_t>(Result) < 0` would still pass if the negative value was too negative to fit in 32 bits, though.

`static_cast<int64_t>(Result) == static_cast<int32_t>(Result)` would work though, and look less ugly

Reply:
`static_cast<int32_t>(Result)` would be undefined behavior (notably breaking ubsan). How about changing `Result`s type to `int64_t` so we can do an explicit range check? `Result < 0x1'00000000 && Result > -0x1'00000000`? Otherwise, I'm guessing your original code will do but it seems rather subtle.

Reply:
Result has to be uint64_t as that's how the interface for `TrackMmap` and `FEX::HLE::HasSyscallError` is defined, unless why add extra casts there.

static cast would be undefined behaviour? I thought it's implementation defined (as C/C++ doesn't explicitly handle negatives as two's complement), similar to (int32_t)(...) would be.

Reply:
Casting large values to `int32_t` is equivalent to signed integer overflow, which is undefined behavior.

Reply:
Fex relies largely on that behaviour being defined, as there are several instances of such casting, and we're also likely to depend on signed integer overflow in other places as well.

The linux kernel also has cases where it depends on such behaviour, at least from the code I've read there so far.

Reply:
Out of curiosity I looked a bit further into this, and [according to this post that suppsoedly quotes the standard](https://stackoverflow.com/questions/67755339/is-signed-integer-overflow-undefined-behaviour-or-implementation-defined) casts are "implementation defined" and signed math overflows is "undefined behaviour".

I guess we should look the exact standard that applies to our case (C++20, it fallbacks to C99 or C11 for these definitions?), though, in practice, every emulator I've worked on depends on these being well behaved, and do what the hardware does.

Also, from the same post, `Some compilers have a command line option to change the behavior of signed arithmetic overflow from undefined behavior to implementation-defined: gcc and clang support -fwrapv` which we probably want to enable.

Reply:
Split follow up #1692